hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -1095,4 +1089,274 @@ public synchronized void close() throws IOException {
       clientKeyStoresFactory.destroy();
     }
   }
+
+  /**
+   * Check how much time before certificate will enter expiry grace period.
+   * @return Duration, time before certificate enters the grace
+   *                   period defined by "hdds.x509.renew.grace.duration"
+   */
+  public Duration timeBeforeExpiryGracePeriod(String certId)
+      throws CertificateException {
+    X509Certificate cert = getCertificate(certId);
+    Duration gracePeriod = securityConfig.getRenewalGracePeriod();
+    Date expireDate = cert.getNotAfter();
+    LocalDateTime gracePeriodStart = expireDate.toInstant()
+        .atZone(ZoneId.systemDefault()).toLocalDateTime().minus(gracePeriod);
+    LocalDateTime currentTime = LocalDateTime.now();
+    if (gracePeriodStart.isBefore(currentTime)) {
+      // Cert is already in grace period time.
+      return Duration.ZERO;
+    } else {
+      return Duration.between(currentTime, gracePeriodStart);
+    }
+  }
+
+  public String renewAndStoreKeyAndCertificate(boolean force)
+      throws CertificateException {
+    if (isRenewing.compareAndSet(false, true)) {
+      try {
+        if (!force) {
+          synchronized (this) {
+            Preconditions.checkArgument(
+                timeBeforeExpiryGracePeriod(certSerialId).isZero());
+          }
+        }
+        String newKeyPath = securityConfig.getKeyLocation(component)
+            .toString() + HDDS_NEW_KEY_CERT_DIR_NAME_SUFFIX;
+        String newCertPath = securityConfig.getCertificateLocation(component)
+            .toString() + HDDS_NEW_KEY_CERT_DIR_NAME_SUFFIX;
+        File newKeyDir = new File(newKeyPath);
+        File newCertDir = new File(newCertPath);
+
+        try {
+          FileUtils.deleteDirectory(newKeyDir);
+          FileUtils.deleteDirectory(newCertDir);
+        } catch (IOException e) {
+          throw new CertificateException("Error while deleting " + newKeyPath +
+              " or " + newCertPath + " directories to cleanup certificate " +
+              " storage. ", e, RENEW_ERROR);
+        }
+
+        try {
+          Files.createDirectories(newKeyDir.toPath());
+          Files.createDirectories(newCertDir.toPath());
+        } catch (IOException e) {
+          throw new CertificateException("Error while creating " + newKeyPath +
+              " or " + newCertPath + " directories for certificate storage.",
+              e, RENEW_ERROR);
+        }
+
+        // cleanup backup directory
+        cleanBackupDir();
+
+        // Generate key
+        KeyCodec newKeyCodec = new KeyCodec(securityConfig, newKeyDir.toPath());
+        KeyPair newKeyPair;
+        try {
+          newKeyPair = createKeyPair(newKeyCodec);
+        } catch (CertificateException e) {
+          throw new CertificateException("Error while creating new key pair.",
+              e, RENEW_ERROR);
+        }
+
+        // Get certificate signed
+        String dnCertSerialId;
```

Comment:
dn prefix I believe is unintentional here.

Reply:
Right.